Src/Workspaces/Core/Portable/Utilities/TaskExtensions.cs

```diff
@@ -337,5 +338,137 @@ public static T WaitAndGetResult<T>(this Task<T> task, CancellationToken cancell
                     _ => continuationFunction(t), cancellationToken, TaskContinuationOptions.None, scheduler),
                 cancellationToken, taskContinuationOptions, scheduler).Unwrap();
         }
+
+        /// <summary>
+        /// Wraps a task with one that will complete as cancelled based on a cancellation token, 
+        /// allowing someone to await a task but be able to break out early by cancelling the token.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by the task.</typeparam>
+        /// <param name="task">The task to wrap.</param>
+        /// <param name="cancellationToken">The token that can be canceled to break out of the await.</param>
+        /// <returns>The wrapping task.</returns>
+        public static Task<T> WithCancellation<T>(this Task<T> task, CancellationToken cancellationToken)
```

Comment:
I'm not sure I'm a fan of this method on principle: it seems it's asking for people to do bad things where they have tasks that they'll cancel their wait on but the task itself was unnecessary. I'd prefer the siren's call not be there if possible. We have a specific problem -- Task.WhenAll not having cancellation. If we write a helper for just that is it a lot smaller?